Add JWT validation endpoint with Sequelize integration

Brightiya opened the pull request:
### 📄 **Pull Request Description**
In response to  [issues 26](https://github.com/ManahilMuhammad/asp-team-20/issues/26)

---

### 📝 **Major Changes**
- JWT validation

---

### ✅ **Type of Modifications**

- 🍃 New Feature  
 

---

### 🔗 **Related Issues**
<!-- Link to any related issues or pull requests, if applicable. Use: Fixes #123 or Closes #456 -->
- Fixes [26](https://github.com/ManahilMuhammad/asp-team-20/issues/26)


---

### 📊 **Testing Steps**
<!-- Outline how your changes can be tested. Include steps, environments, or edge cases considered. -->
1.  cd  src/server/
2. Run npm install
3.  To launch the server and verify the route  enter node server.js . Visit  http://localhost:3000/auth/validate

---

### 🛑 **Checklist Before Merging**
- [x] Code follows the project's coding standards.
- [x] Appropriate tests have been added or updated.
- [ ] Documentation updated where necessary.
- [x] PR title and description are clear and concise.
- [x] All checks pass (CI/CD, linting).

---

### 💬 **Additional Context**
Non

Brightiya committed:
Add JWT validation endpoint with Sequelize integration

Maximus7474 commented on src/server/middleware/auth.middelware.js:
Won't this cause issues when using any other protected endpoint ?

Maximus7474 replied:
Update: So it does work, thought it might cause issues with the `res.json({ valid: true });`

I'm still on the opinion on keeping the middleware used to protect endpoints separate from the JWT token validation function.
Maybe export the protect function and a separate JWT verification ?

Brightiya replied:
Yes, done.

Maximus7474 commented on src/server/routes/auth.js:
Best to use POST in this case I'd say ?

Maximus7474 commented:
Unable to test currently as I'm on my phone. Just the two comments on the PR to clear up and I think it should be good.

Brightiya commented:
Well spotted. I’ll  make an update once back at home. Thanks.

On Tue 25 Feb 2025 at 12:02, Maximus7474 ***@***.***> wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> On src/server/middleware/auth.middelware.js
> <https://github.com/ManahilMuhammad/asp-team-20/pull/27#discussion_r1969542628>
> :
>
> Update: So it does work, thought it might cause issues with the res.json({
> valid: true });
>
> I'm still on the opinion on keeping the middleware used to protect
> endpoints separate from the JWT token validation function.
> Maybe export the protect function and a separate JWT verification ?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/ManahilMuhammad/asp-team-20/pull/27#discussion_r1969542628>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ARMQSX3JVINJYNDWS3C73XL2RRETVAVCNFSM6AAAAABXZRMIAOVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZDMNBQGYZTINBWGQ>
> .
> You are receiving this because you were assigned.Message ID:
> ***@***.***>
>

Maximus7474 commented:
I started working on the auth verification on the client side, running into a couple of issues.
Will add it to this branch as it's the same scope

Maximus7474 commented:
@Brightiya Found my issue I didn't check how the JWT was built serverside, I blindly assumed that it would also contain email and name for the user.

So for the `api/auth/validate` endpoint maybe alter it so it returns the user data to update it on the client, this will allow us to make sure that the "general" details of the user clientside are synced with the server.
We will be able to integrate (later on) things such as profile picture so it can be displayed.

Maximus7474 committed:
refactor(client/auth-provider): query server endpoint to validate token
-> will be functional when endpoint is set to POST & not GET
-> will be tweaked to accept along with the server token user details (email, name, profile pic, etc...)

Maximus7474 committed:
feat(client/components): protected-route
To Do:
-> add a loading screen until user is verified, as shown on slack - https://londoncs.slack.com/archives/C080KMTQUM7/p1740163504422879?thread_ts=1740059370.338269&cid=C080KMTQUM7

Maximus7474 committed:
chore(client/App): implement protected route for profile page

Brightiya committed:
refactor/chore: /feature/jwt-validation

Brightiya committed:
Merge remote-tracking branch 'refs/remotes/origin/feature/jwt-validation' into feature/jwt-validation

Brightiya committed:
add protected-route.tsx to local main

Brightiya commented:
@Maximus7474  I update the code according to the issue you  correctly pointed out.  For  the point you raised for the user data, I think the user data is already being sent to the client-side when the user login.  I saw  from _auth-provider.tsx_ you already implemented localStorage  : `if (!response.ok) {
        throw new Error("Invalid credentials");
      }
const { token } = await response.json();
      localStorage.setItem("nutrifit-token", token);`, so you could get the user data from the response like so?:
      `localStorage.setItem('userData', JSON.stringify(response.user)` and then: `const user = localStorage.getItem('userData'); 
      return user ? JSON.parse(user) : null;`

Maximus7474 commented:
> I saw from _auth-provider.tsx_ you already implemented localStorage : `if (!response.ok) { throw new Error("Invalid credentials"); } const { token } = await response.json(); localStorage.setItem("nutrifit-token", token);`, so you could get the user data from the response like so?: `localStorage.setItem('userData', JSON.stringify(response.user)` and then: `const user = localStorage.getItem('userData'); return user ? JSON.parse(user) : null;`

So you suggest saving the userData locally ? 
The issue with this is that the name, profile pic, etc... won't be synced across devices. 

I'd suggest doing something like:
```js
  try {
    const decoded = jwt.verify(token, JWT_SECRET);
    const user = await db.User.findByPk(decoded.id);
    res.json({ valid: true, user });
  } catch (error) {
    res.status(401).json({ valid: false, message: "Invalid or expired token" });
  }
```

Brightiya commented:
Brilliant! Okay.
Thanks

On Wed 26 Feb 2025 at 10:55, Maximus7474 ***@***.***> wrote:

> I saw from *auth-provider.tsx* you already implemented localStorage : if
> (!response.ok) { throw new Error("Invalid credentials"); } const { token }
> = await response.json(); localStorage.setItem("nutrifit-token", token);,
> so you could get the user data from the response like so?: localStorage.setItem('userData',
> JSON.stringify(response.user) and then: const user =
> localStorage.getItem('userData'); return user ? JSON.parse(user) : null;
>
> So you suggest saving the userData locally ?
> The issue with this is that the name, profile pic, etc... won't be synced
> across devices.
>
> I'd suggest doing something like:
>
>   try {
>     const decoded = jwt.verify(token, JWT_SECRET);
>     const user = await db.User.findByPk(decoded.id);
>     res.json({ valid: true, user });
>   } catch (error) {
>     res.status(401).json({ valid: false, message: "Invalid or expired token" });
>   }
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/ManahilMuhammad/asp-team-20/pull/27#issuecomment-2684470344>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ARMQSX4FZAN4R44GSJZTXPD2RWFQRAVCNFSM6AAAAABXZRMIAOVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDMOBUGQ3TAMZUGQ>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
> [image: Maximus7474]*Maximus7474* left a comment
> (ManahilMuhammad/asp-team-20#27)
> <https://github.com/ManahilMuhammad/asp-team-20/pull/27#issuecomment-2684470344>
>
> I saw from *auth-provider.tsx* you already implemented localStorage : if
> (!response.ok) { throw new Error("Invalid credentials"); } const { token }
> = await response.json(); localStorage.setItem("nutrifit-token", token);,
> so you could get the user data from the response like so?: localStorage.setItem('userData',
> JSON.stringify(response.user) and then: const user =
> localStorage.getItem('userData'); return user ? JSON.parse(user) : null;
>
> So you suggest saving the userData locally ?
> The issue with this is that the name, profile pic, etc... won't be synced
> across devices.
>
> I'd suggest doing something like:
>
>   try {
>     const decoded = jwt.verify(token, JWT_SECRET);
>     const user = await db.User.findByPk(decoded.id);
>     res.json({ valid: true, user });
>   } catch (error) {
>     res.status(401).json({ valid: false, message: "Invalid or expired token" });
>   }
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/ManahilMuhammad/asp-team-20/pull/27#issuecomment-2684470344>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ARMQSX4FZAN4R44GSJZTXPD2RWFQRAVCNFSM6AAAAABXZRMIAOVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDMOBUGQ3TAMZUGQ>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

Maximus7474 committed:
tweak(client/auth-provider): adapted endpoint to work with commit ec6692c49b5cce78254cab7859ca0ab980ee1e0e

Maximus7474 committed:
fix(server/routes/auth): set as a POST request and not GET

Maximus7474 committed:
refactor(auth/token-validation): send user data to client to ensure proper data sync

Maximus7474 committed:
tweak(server/controller/user): alter response body structure

Maximus7474 committed:
fix(client/auth-provider): store user data properly for the session

Maximus7474 committed:
chore(client/auth-provider): comment out parseJWT function - no longer needed

Maximus7474 commented:
I've added the discussed changes and refactored some parts to better work with previous changes.

Brightiya commented:
@good to merge now?

Brightiya commented on src/server/routes/auth.js:
Just verified, router.POST generated a 'a cannot GET ' error, which is expected because this is a 'Read-Only Operation'. The endpoint only reads data (the token from the header) and returns a validation result. It doesn't create or update any resource or data. So,  I changed it back to GET, then it worked.

Maximus7474 replied:
That'll break the auth-provider token check which processes the request as a `POST` req.
It avoids people spamming the page through the browser, people can still use curl or custom scripts to do so but it's more secure and a general practice for anything auth related to use POST, there's also no character limit on these compared to GET requests

Maximus7474 replied:
So the decision to take is do we stick with the latest change of using `GET` requiring it to be changed client side or do we revert back to `POST`

Brightiya committed:
refactor: auth.js route

Brightiya commented:
Make sense, let’s revert back to POST then. You can merge now if everything
works?

On Fri 28 Feb 2025 at 11:58, Maximus7474 ***@***.***> wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> On src/server/routes/auth.js
> <https://github.com/ManahilMuhammad/asp-team-20/pull/27#discussion_r1975218995>
> :
>
> So the decision to take is do we stick with the latest change of using GET
> requiring it to be changed client side or do we revert back to POST
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/ManahilMuhammad/asp-team-20/pull/27#discussion_r1975218995>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ARMQSX2IQRVXR2GNVSONVXT2SA6MBAVCNFSM6AAAAABXZRMIAOVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZDMNJQGQZDIMRYGM>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

Maximus7474 commented:
> Make sense, let’s revert back to POST then. You can merge now if everything works?


Sounds good will do

Maximus7474 committed:
revert: refactor: auth.js route
